File: example/scalalib/web/3-todo-http4s/src/WebApp.scala
package webapp

import scala.concurrent.duration.Duration
import scalatags.Text.all.*
import scalatags.Text.tags2
import cats.effect.*
import cats.syntax.all.*
import com.comcast.ip4s.*
import org.http4s.*
import org.http4s.dsl.io.*
import org.http4s.ember.server.*
import org.http4s.scalatags.*
import org.http4s.server.staticcontent.*
import io.circe.*
import io.circe.generic.semiauto.*

object WebApp extends IOApp.Simple {
  case class Todo(checked: Boolean, text: String)

  def run = mkService.toResource.flatMap { service =>
    EmberServerBuilder
      .default[IO]
      .withHttpApp(service)
      .withPort(port"8084")
      .withShutdownTimeout(Duration.Zero)
      .build
  }.useForever

  def mkService =
    IO.ref(Seq(Todo(true, "Get started with http4s"), Todo(false, "Profit!"))).map { todosRef =>
      def apiRoutes = HttpRoutes.of[IO] {
        case POST -> Root / "list" / state =>
          Ok(renderBody(state))

        case request @ POST -> Root / "add" / state =>
          for {
            text <- request.as[String]
            _ <- todosRef.update(Seq(Todo(false, text)) ++ _)
            response <- Ok(renderBody(state))
          } yield response

        case POST -> Root / "delete" / state / IntVar(index) =>
          todosRef.update(_.patch(index, Nil, 1)) *> Ok(renderBody(state))

        case POST -> Root / "toggle" / state / IntVar(index) =>
          todosRef.update { todos =>
            todos.updated(index, todos(index).copy(checked = !todos(index).checked))
          } *> Ok(renderBody(state))

        case POST -> Root / "clear-completed" / state =>
          todosRef.update(_.filter(!_.checked)) *> Ok(renderBody(state))

        case POST -> Root / "toggle-all" / state =>
          todosRef.update { todos =>
            val next = todos.filter(_.checked).size != 0
            todos.map(_.copy(checked = next))
          } *> Ok(renderBody(state))

        case GET -> Root => Ok(index)
      }

      def renderBody(state: String) =
        todosRef.get.map { todos =>
          val filteredTodos = state match {
            case "all" => todos.zipWithIndex
            case "active" => todos.zipWithIndex.filter(!_._1.checked)
            case "completed" => todos.zipWithIndex.filter(_._1.checked)
          }
          div(
            header(
              cls := "header",
              h1("todos"),
              input(cls := "new-todo", placeholder := "What needs to be done?", autofocus := "")
            ),
            tags2.section(
              cls := "main",
              input(
                id := "toggle-all",
                cls := "toggle-all",
                `type` := "checkbox",
                if (todos.filter(_.checked).size != 0) checked else ()
              ),
              label(`for` := "toggle-all", "Mark all as complete"),
              ul(
                cls := "todo-list",
                for ((todo, index) <- filteredTodos) yield li(
                  if (todo.checked) cls := "completed" else (),
                  div(
                    cls := "view",
                    input(
                      cls := "toggle",
                      `type` := "checkbox",
                      if (todo.checked) checked else (),
                      data("todo-index") := index
                    ),
                    label(todo.text),
                    button(cls := "destroy", data("todo-index") := index)
                  ),
                  input(cls := "edit", value := todo.text)
                )
              )
            ),
            footer(
              cls := "footer",
              span(cls := "todo-count", strong(todos.filter(!_.checked).size), " items left"),
              ul(
                cls := "filters",
                li(cls := "todo-all", a(if (state == "all") cls := "selected" else (), "All")),
                li(
                  cls := "todo-active",
                  a(if (state == "active") cls := "selected" else (), "Active")
                ),
                li(
                  cls := "todo-completed",
                  a(if (state == "completed") cls := "selected" else (), "Completed")
                )
              ),
              button(cls := "clear-completed", "Clear completed")
            )
          )
        }

      def index = renderBody("all").map { renderedBody =>
        doctype("html")(
          html(
            lang := "en",
            head(
              meta(charset := "utf-8"),
              meta(name := "viewport", content := "width=device-width, initial-scale=1"),
              tags2.title("Template • TodoMVC"),
              link(rel := "stylesheet", href := "/static/index.css")
            ),
            body(
              tags2.section(cls := "todoapp", renderedBody),
              footer(
                cls := "info",
                p("Double-click to edit a todo"),
                p("Created by ", a(href := "http://todomvc.com", "Li Haoyi")),
                p("Part of ", a(href := "http://todomvc.com", "TodoMVC"))
              ),
              script(src := "/static/main.js")
            )
          )
        )
      }

      def staticRoutes = resourceServiceBuilder[IO]("webapp").withPathPrefix("static").toRoutes

      (apiRoutes <+> staticRoutes).orNotFound
    }

}

File: example/kotlinlib/web/5-webapp-kotlinjs/src/WebApp.kt
package webapp

import io.ktor.server.application.Application
import io.ktor.server.application.call
import io.ktor.server.engine.embeddedServer
import io.ktor.server.html.respondHtml
import io.ktor.server.http.content.staticResources
import io.ktor.server.netty.Netty
import io.ktor.server.request.receiveText
import io.ktor.server.response.respondText
import io.ktor.server.routing.get
import io.ktor.server.routing.post
import io.ktor.server.routing.routing
import io.ktor.server.util.getOrFail
import kotlinx.html.FlowContent
import kotlinx.html.HTML
import kotlinx.html.InputType
import kotlinx.html.a
import kotlinx.html.body
import kotlinx.html.button
import kotlinx.html.div
import kotlinx.html.footer
import kotlinx.html.form
import kotlinx.html.h1
import kotlinx.html.head
import kotlinx.html.header
import kotlinx.html.id
import kotlinx.html.input
import kotlinx.html.label
import kotlinx.html.li
import kotlinx.html.link
import kotlinx.html.meta
import kotlinx.html.p
import kotlinx.html.script
import kotlinx.html.section
import kotlinx.html.span
import kotlinx.html.stream.createHTML
import kotlinx.html.strong
import kotlinx.html.title
import kotlinx.html.ul

object WebApp {
    data class Todo(
        val checked: Boolean,
        val text: String,
    )

    private val todos =
        mutableListOf(
            Todo(true, "Get started with Cask"),
            Todo(false, "Profit!"),
        )

    private fun FlowContent.list(state: String) = renderBody(state)

    private fun FlowContent.add(
        state: String,
        text: String,
    ) {
        todos.add(Todo(false, text))
        renderBody(state)
    }

    private fun FlowContent.delete(
        state: String,
        index: Int,
    ) {
        todos.removeAt(index)
        renderBody(state)
    }

    private fun FlowContent.toggle(
        state: String,
        index: Int,
    ) {
        todos[index] =
            todos[index].let {
                it.copy(checked = !it.checked)
            }
        renderBody(state)
    }

    private fun FlowContent.clearCompleted(state: String) {
        todos.removeAll { it.checked }
        renderBody(state)
    }

    private fun FlowContent.toggleAll(state: String) {
        val next = todos.any { !it.checked }
        for (item in todos.withIndex()) {
            todos[item.index] = item.value.copy(checked = next)
        }
        renderBody(state)
    }

    private fun FlowContent.renderBody(state: String) {
        val filteredTodos =
            when (state) {
                "all" -> todos.withIndex()
                "active" -> todos.withIndex().filter { !it.value.checked }
                "completed" -> todos.withIndex().filter { it.value.checked }
                else -> throw IllegalStateException("Unknown state=$state")
            }
        div {
            header(classes = "header") {
                h1 {
                    +"todos"
                }
                input(classes = "new-todo") {
                    placeholder = "What needs to be done?"
                }
            }
            section(classes = "main") {
                input(
                    classes = "toggle-all",
                    type = InputType.checkBox,
                ) {
                    id = "toggle-all"
                    checked = todos.any { it.checked }
                }
                label {
                    htmlFor = "toggle-all"
                    +"Mark all as complete"
                }
                ul(classes = "todo-list") {
                    filteredTodos.forEach { (index, todo) ->
                        li(classes = if (todo.checked) "completed" else "") {
                            div(classes = "view") {
                                form {
                                    input(classes = "toggle", type = InputType.checkBox) {
                                        checked = todo.checked
                                        attributes["data-todo-index"] = index.toString()
                                    }
                                    label { +todo.text }
                                }
                                form {
                                    button(classes = "destroy") {
                                        attributes["data-todo-index"] = index.toString()
                                    }
                                }
                            }
                            input(classes = "edit") {
                                value = todo.text
                            }
                        }
                    }
                }
            }
            footer(classes = "footer") {
                span(classes = "todo-count") {
                    strong {
                        +todos.filter { !it.checked }.size.toString()
                    }
                    +" items left"
                }
                ul(classes = "filters") {
                    li(classes = "todo-all") {
                        a(classes = if (state == "all") "selected" else "") { +"All" }
                    }
                    li(classes = "todo-active") {
                        a(classes = if (state == "active") "selected" else "") { +"Active" }
                    }
                    li(classes = "todo-completed") {
                        a(classes = if (state == "completed") "selected" else "") { +"Completed" }
                    }
                }
                button(classes = "clear-completed") { +"Clear completed" }
            }
        }
    }

    private fun HTML.renderIndex() {
        head {
            meta(charset = "utf-8")
            meta(name = "viewport", content = "width=device-width, initial-scale=1")
            title("Template • TodoMVC")
            link(rel = "stylesheet", href = "/static/index.css")
        }
        body {
            section(classes = "todoapp") {
                renderBody("all")
            }
            footer(classes = "info") {
                p { +"Double-click to edit a todo" }
                p {
                    +"Created by "
                    a(href = "http://todomvc.com") { +"Li Haoyi" }
                }
                p {
                    +"Part of "
                    a(href = "http://todomvc.com") { +"TodoMVC" }
                }
            }
            script(src = "/static/client.js", block = {})
        }
    }

    fun configureRoutes(app: Application) {
        with(app) {
            routing {
                get("/") {
                    call.respondHtml {
                        renderIndex()
                    }
                }
                post("/toggle-all/{state}") {
                    call.respondText {
                        createHTML().div { toggleAll(call.parameters.getOrFail("state")) }
                    }
                }
                post("/clear-completed/{state}") {
                    call.respondText {
                        createHTML().div { clearCompleted(call.parameters.getOrFail("state")) }
                    }
                }
                post("/toggle/{state}/{index}") {
                    call.parameters.run {
                        call.respondText {
                            createHTML().div { toggle(getOrFail("state"), getOrFail<Int>("index")) }
                        }
                    }
                }
                post("/delete/{state}/{index}") {
                    call.parameters.run {
                        call.respondText {
                            createHTML().div { delete(getOrFail("state"), getOrFail<Int>("index")) }
                        }
                    }
                }
                post("/add/{state}") {
                    val requestText = call.receiveText()
                    call.respondText {
                        createHTML().div { add(call.parameters.getOrFail("state"), requestText) }
                    }
                }
                post("/list/{state}") {
                    call.respondText {
                        createHTML().div { list(call.parameters.getOrFail("state")) }
                    }
                }
                staticResources("/static", "webapp")
            }
        }
    }

    @JvmStatic
    fun main(args: Array<String>) {
        embeddedServer(Netty, port = 8092, host = "0.0.0.0") {
            configureRoutes(this)
        }.start(wait = true)
    }
}
